### PR TITLE
Issue 30 serverless transform

### PR DESCRIPTION
**Pull request by jeffb4**
This adds a generic transform modeling capability, and specifically a SAM serverless transform

Tests not yet completed, putting out early for pattern review

### CODE REVIEW

**Comment by jeffb4** [comment]
Tests added, pipelines successfully

File: lib/cfn-model/parser/cfn_parser.rb [hunk 1]
```diff
@@ -31,7 +31,13 @@ class CfnParser
   # Given raw json/yml CloudFormation template, returns a CfnModel object
   # or raise ParserErrors if something is amiss with the format
   def parse(cloudformation_yml, parameter_values_json=nil)
-    cfn_hash = pre_validate_model cloudformation_yml
```
Comment (erickascic) [inline lib/cfn-model/parser/cfn_parser.rb]: the validator finds badness and then returns the cfn_hash.  what is it the TransformerRegistry needs to access before the yml/json has been "pre-validated"?

File: lib/cfn-model/parser/transform_registry.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+class CfnModel
+  # TransformRegistry provides a registry of CloudFormation transforms
+  # available for templates
+  class TransformRegistry
+    attr_reader :registry
+
+    def initialize
+      @registry = {
+        'AWS::Serverless-2016-10-31' => CfnModel::Transforms::Serverless
```
Comment (erickascic) [inline lib/cfn-model/parser/transform_registry.rb]: remind me to talk about this one offline

File: lib/cfn-model/transforms/serverless.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
+class CfnModel
+  class Transforms
+    # Handle transformation of model elements performed by the
+    # Serverless trasnform, see
+    # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html
+    class Serverless
+      def perform_transform(cfn_hash)
+        resources = cfn_hash['Resources'].clone
+        resources.each do |r_name, r_value|
```
Comment (erickascic) [inline lib/cfn-model/transforms/serverless.rb]: i prefer |resource_name, resource| here

File: lib/cfn-model/transforms/serverless.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          next unless r_value['Type'].eql? 'AWS::Serverless::Function'
```
Comment (erickascic) [inline lib/cfn-model/transforms/serverless.rb]: there are a few beyond the Function: https://awslabs.github.io/serverless-application-model/internals/generated_resources.html

Comment (erickascic) [reply lib/cfn-model/transforms/serverless.rb]: perhaps a mapping here between serverless resources and the replace_* method?

File: lib/cfn-model/transforms/serverless.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        @instance ||= Serverless.new
+        @instance
+      end
+
+      private
+
+      def replace_serverless_function(cfn_hash, resource_name)
+        resource = cfn_hash['Resources'][resource_name]
+        # Bucket is 3rd element of an S3 URI split on '/'
+        code_bucket = \
```
Comment (erickascic) [inline lib/cfn-model/transforms/serverless.rb]: the \ drives my eyes nuts in RubyMine, but that aside - suggest a named method to encapsulate the extraction of the token/element

File: lib/cfn-model/transforms/serverless.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        cfn_hash['Resources'][resource_name] = \
+          lambda_function(
+            handler: resource['Properties']['Handler'],
+            code_bucket: code_bucket,
+            code_key: code_key,
+            runtime: resource['Properties']['Runtime']
+          )
+
+        cfn_hash['Resources']['FunctionNameRole'] = function_name_role
+        puts "cfn_hash = #{cfn_hash}"
```
Comment (erickascic) [inline lib/cfn-model/transforms/serverless.rb]: no useful return value here?  even just for testing....?

File: lib/cfn-model/parser/cfn_parser.rb (hunk 1 above)
Comment (jeffb4) [reply lib/cfn-model/parser/cfn_parser.rb]: It needs to create / modify resources in the YAML hash before higher level validation (like references being defined) can be successfully performed

File: lib/cfn-model/parser/transform_registry.rb (hunk 2 above)
Comment (jeffb4) [reply lib/cfn-model/parser/transform_registry.rb]: (done)

**Comment by jeffb4** [comment]
One thing that got lost in the code review was the comment that we don't create all of the resources defined by the https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md transform doc.

We don't, but I'd like a start we can build off of.

**Pull request merged**